### PR TITLE
[4.0][webservice][com_actionlogs] - actionlogs api endpoint

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
added an api endpoint for com_actionlogs

### Testing Instructions
if you use postman 
then `GET http://localhost/<yoursite>/api/index.php/v1/actionlogs`

`GET http://localhost/<yoursite>/api/index.php/v1/actionlogs/xx`

where xx is the log id

`GET http://localhost/<yoursite>/api/index.php/v1/actionlogs?extension=com_xxxx`

where xxxx is the extension name

`GET http://localhost/<yoursite>/api/index.php/v1/actionlogs?user=xxxx`

where xxxx is the user id

if you use curl 
`curl -X GET /api/index.php/v1/actionlogs`

### Expected result
something like 
```json
{
    "links": {
        "self": "http://localhost/my4/api/index.php/v1/actionlogs",
        "first": "http://localhost/my4/api/index.php/v1/actionlogs?page[offset]=0&page[limit]=20",
        "next": "http://localhost/my4/api/index.php/v1/actionlogs?page[offset]=20&page[limit]=20",
        "previous": "http://localhost/my4/api/index.php/v1/actionlogs?page[offset]=0&page[limit]=20",
        "last": "http://localhost/my4/api/index.php/v1/actionlogs?page[offset]=0&page[limit]=20"
    },
    "data": [
        {
            "type": "actionlogs",
            "id": "14",
            "attributes": {
                "id": "14",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_API\"}",
                "log_date": "2019-10-28 19:14:54",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "13",
            "attributes": {
                "id": "13",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_API\"}",
                "log_date": "2019-10-28 19:14:15",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "12",
            "attributes": {
                "id": "12",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_API\"}",
                "log_date": "2019-10-28 19:13:21",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "11",
            "attributes": {
                "id": "11",
                "message_language_key": "PLG_SYSTEM_ACTIONLOGS_CONTENT_UPDATED",
                "message": "{\"action\":\"update\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_COMPONENT_CONFIG\",\"id\":\"213\",\"title\":\"com_patchtester\",\"extension_name\":\"com_patchtester\",\"itemlink\":\"index.php?option=com_config&task=component.edit&extension_id=213\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:48:12",
                "extension": "com_config.component",
                "user_id": "607",
                "item_id": "213",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "10",
            "attributes": {
                "id": "10",
                "message_language_key": "PLG_SYSTEM_ACTIONLOGS_CONTENT_UPDATED",
                "message": "{\"action\":\"update\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_COMPONENT_CONFIG\",\"id\":\"213\",\"title\":\"com_patchtester\",\"extension_name\":\"com_patchtester\",\"itemlink\":\"index.php?option=com_config&task=component.edit&extension_id=213\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:47:39",
                "extension": "com_config.component",
                "user_id": "607",
                "item_id": "213",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "9",
            "attributes": {
                "id": "9",
                "message_language_key": "PLG_SYSTEM_ACTIONLOGS_CONTENT_UPDATED",
                "message": "{\"action\":\"update\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_COMPONENT_CONFIG\",\"id\":\"213\",\"title\":\"com_patchtester\",\"extension_name\":\"com_patchtester\",\"itemlink\":\"index.php?option=com_config&task=component.edit&extension_id=213\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:45:57",
                "extension": "com_config.component",
                "user_id": "607",
                "item_id": "213",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "8",
            "attributes": {
                "id": "8",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_EXTENSION_INSTALLED",
                "message": "{\"action\":\"install\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_COMPONENT\",\"id\":\"213\",\"name\":\"com_patchtester\",\"extension_name\":\"com_patchtester\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:44:12",
                "extension": "com_installer",
                "user_id": "607",
                "item_id": "213",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "7",
            "attributes": {
                "id": "7",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_CHECKIN",
                "message": "{\"action\":\"checkin\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_USER\",\"id\":\"607\",\"title\":\"admin\",\"itemlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"table\":\"#__extensions\"}",
                "log_date": "2019-10-28 18:41:52",
                "extension": "com_checkin",
                "user_id": "607",
                "item_id": "607",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "6",
            "attributes": {
                "id": "6",
                "message_language_key": "PLG_SYSTEM_ACTIONLOGS_CONTENT_UPDATED",
                "message": "{\"action\":\"update\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_PLUGIN\",\"id\":214,\"title\":\"plg_webservices_actionlog\",\"extension_name\":\"plg_webservices_actionlog\",\"itemlink\":\"index.php?option=com_plugins&task=plugin.edit&extension_id=214\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:41:52",
                "extension": "com_plugins.plugin",
                "user_id": "607",
                "item_id": "214",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "5",
            "attributes": {
                "id": "5",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_EXTENSION_INSTALLED",
                "message": "{\"action\":\"install\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_PLUGIN\",\"id\":\"214\",\"name\":\"plg_webservices_actionlog\",\"extension_name\":\"plg_webservices_actionlog\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:41:24",
                "extension": "com_installer",
                "user_id": "607",
                "item_id": "214",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "4",
            "attributes": {
                "id": "4",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_API\"}",
                "log_date": "2019-10-28 18:28:24",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "3",
            "attributes": {
                "id": "3",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_API\"}",
                "log_date": "2019-10-28 18:27:57",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "2",
            "attributes": {
                "id": "2",
                "message_language_key": "PLG_SYSTEM_ACTIONLOGS_CONTENT_ADDED",
                "message": "{\"action\":\"add\",\"type\":\"PLG_ACTIONLOG_JOOMLA_TYPE_ARTICLE\",\"id\":\"1\",\"title\":\"AAAAAA\",\"itemlink\":\"index.php?option=com_content&task=article.edit&id=1\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\"}",
                "log_date": "2019-10-28 18:27:50",
                "extension": "com_content.article",
                "user_id": "607",
                "item_id": "1",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        },
        {
            "type": "actionlogs",
            "id": "1",
            "attributes": {
                "id": "1",
                "message_language_key": "PLG_ACTIONLOG_JOOMLA_USER_LOGGED_IN",
                "message": "{\"action\":\"login\",\"userid\":\"607\",\"username\":\"admin\",\"accountlink\":\"index.php?option=com_users&task=user.edit&id=607\",\"app\":\"PLG_ACTIONLOG_JOOMLA_APPLICATION_ADMINISTRATOR\"}",
                "log_date": "2019-10-28 18:27:37",
                "extension": "com_users",
                "user_id": "607",
                "item_id": "0",
                "ip_address": "COM_ACTIONLOGS_DISABLED"
            }
        }
    ],
    "meta": {
        "total-pages": 1
    }
}
```

### Actual result
N\A

### Documentation Changes Required
yes

### CODE REVIEW

File: plugins/webservices/actionlogs/actionlog.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\CMS\Router\ApiRouter;
+use Joomla\Router\Route;
+
+/**
+ * Web Services adapter for com_content.
```
Comment (brianteeman) [inline plugins/webservices/actionlogs/actionlog.php]:
```suggestion
 * Web Services adapter for com_actionlogs
```

File: api/components/com_actionlogs/View/Actionlogs/JsonapiView.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+use Joomla\CMS\MVC\View\GenericDataException;
+use Joomla\CMS\MVC\View\JsonApiView as BaseApiView;
+use Joomla\CMS\Router\Exception\RouteNotFoundException;
+use Joomla\CMS\Serializer\JoomlaSerializer;
+use Joomla\CMS\Uri\Uri;
+use Tobscure\JsonApi\Resource;
+
+/**
+ * The consents view
```
Comment (brianteeman) [inline api/components/com_actionlogs/View/Actionlogs/JsonapiView.php]: ??

Comment (alikon) [reply api/components/com_actionlogs/View/Actionlogs/JsonapiView.php]: a c&p laziness :face_with_head_bandage:

**Comment by Quy** [comment]
Installation failed.

> PHP Warning:  simplexml_load_file(): I/O warning : failed to load external entity &quot;file:/joomla-cms-patch-111/plugins/webservices/actionlog/actionlog.xml&quot; in \libraries\src\Installer\Installer.php on line 2034

**Comment by Quy** [comment]
Change the path to `plugins/webservices/actionlog/` without `s`  for the 2 files fixed the installation issue.

**Comment by Quy** [comment]
In User Actions Log: `User admin logged in to PLG_ACTIONLOG_JOOMLA_APPLICATION_API`
With PR #27094 installed, the above entry is not logged.

**Comment by alikon** [comment]
that's expected by #27094, it logs only performed "**_verb_**" 
p.s
i'll look into the '**s**' issue asap

File: administrator/components/com_actionlogs/Model/ActionlogModel.php [hunk 3]
```diff
@@ -171,4 +172,28 @@ protected function sendNotificationEmails($messages, $username, $context)
...
+	public function getItem($pk = null)
+	{
+		$pk = (!empty($pk)) ? $pk : (int) $this->getState('actionlog.id');
+
+		$db    = $this->getDbo();
+		$query = $db->getQuery(true);
+		$query->select('*')
+			->from($db->quoteName('#__action_logs'))
+			->where($db->quoteName('id') . ' = :id')
+			->bind(':id', $pk, );
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogModel.php]:
```suggestion
			->bind(':id', $pk, ParameterType::INTEGER);
```

File: administrator/components/com_actionlogs/Model/ActionlogModel.php [hunk 4]
```diff
@@ -171,4 +172,28 @@ protected function sendNotificationEmails($messages, $username, $context)
...
+
+		$db->setQuery($query);
+		return $db->loadObject();
```
Comment (Quy) [inline administrator/components/com_actionlogs/Model/ActionlogModel.php:198]:
```suggestion

		return $db->loadObject();
```

**Pull request closed without merging**